cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -225,9 +227,15 @@ public SparkExecutionContext create(SparkRuntimeContext runtimeContext) {
 
         // Localize the hConf file to executor nodes
         localizeResources.add(new LocalizeResource(saveHConf(hConf, tempDir)));
+
+        for (File jar : CConfigurationUtil.getExtraJars(cConf)) {
+          String jarName = jar.getName();
+          extraJars.add(jarName);
+          localizeResources.add(new LocalizeResource(new File(jarName)));
```

Comment:
Seems like you recent changes have my previous comment hidden. You should modify cconf in the `saveCConf` method in `AbstractDistributedProgramRunner` so that when `getExtraJars` is called, it always give the correct location of the list of `File` that it returns.